### PR TITLE
Improvement: Improve rotate canvas camera

### PR DESCRIPTION
**Pull request by bbaa77770**
[Set Rotation Center](https://snapmaker2.atlassian.net/browse/LU-1724)

### CODE REVIEW

File: src/app/ui/components/SMCanvas/Canvas.jsx [hunk 1]
```diff
@@ -60,7 +60,8 @@ class Canvas extends Component {
         showContextMenu: PropTypes.func,
 
         // inProgress
-        inProgress: PropTypes.bool
+        inProgress: PropTypes.bool,
+        headType: PropTypes.string
```
Comment (parachvte) [inline src/app/ui/components/SMCanvas/Canvas.jsx:64]: Don't use `headType` in Canvas.

File: src/app/ui/components/SMCanvas/Controls.jsx [hunk 2]
```diff
@@ -185,11 +192,42 @@ class Controls extends EventEmitter {
         this.sphericalDelta.phi -= angle;
     }
 
+    _getCameraPositionByRotation(positionStart, target, angleEW, angleNS) {
```
Comment (parachvte) [inline src/app/ui/components/SMCanvas/Controls.jsx:195]: Add docs to describe what it is.

File: src/app/ui/components/SMCanvas/Controls.jsx [hunk 3]
```diff
@@ -185,11 +192,42 @@ class Controls extends EventEmitter {
...
+
+            this.target.set(newTarget.x, newTarget.y, newTarget.z);
+            this.camera.position.x = newCameraPosition.x;
+            this.camera.position.y = newCameraPosition.y;
+            this.camera.position.z = newCameraPosition.z;
+        } else {
+            this.rotateLeft(2 * Math.PI * deltaX / elem.clientHeight); // yes, height
+        }
         this.rotateUp(2 * Math.PI * deltaY / elem.clientHeight);
+        this.updateCamera();
```
Comment (parachvte) [inline src/app/ui/components/SMCanvas/Controls.jsx:230]: Why call `this.updateCamera()`? It shouldn't be called frequently.

**Review by parachvte — CHANGES_REQUESTED** [review]
Don't break basic code rules.

File: src/app/ui/components/SMCanvas/Canvas.jsx (hunk 1 above)
Comment (meiliang1123) [reply src/app/ui/components/SMCanvas/Canvas.jsx:64]: +1

**Review by meiliang1123 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**